File: src/ProjectR/content/scripts/afflictions/Debuffs/Paralyze.lua
function GetName()
  return "Paralyze"
end

function GetType()
  return Debuff
end

function GetHookPoints()
  return { TurnCounterUpdating }
end

function OnAttachment(char)
  local charPAR = GetStats(char):GetTotalStat(PAR) * 3

  if charPAR < 20 then
    charPAR = 20 -- We don't want to PAR them forever if they have low Resi, max 5 turns
  end

  local parStrength = GetTimeToAction() * (100 / charPAR);

  char:SetVar("par_Strength", parStrength)
end

function OnRemoval(char)
end

function Trigger(hookPoint, char)
  if hookPoint ~= TurnCounterUpdating then
    return
  end

  local timeStep = GetStats(char):GetTotalStat(SPD)
  local currStrength = char:GetVar("par_Strength")
  local newStrength = currStrength - timeStep
  char:SetVar("par_Strength", newStrength)

  if newStrength <= 0 then
    this:RemoveFrom(char)
  end
  return false
end

File: src/ProjectR.Scripting/content/scripts/afflictions/Debuffs/Paralyze.cs
using ProjectR.Interfaces.Model;
using ProjectR.Interfaces.Model.Stats;

namespace ProjectR.Scripting.Afflictions
{
    public class Paralyze : Affliction
    {
        public override string Name { get { return "Paralyze"; } }

        public override AfflictionType Type { get { return AfflictionType.Debuff; } }

        protected override HookPoint[] HookPoints
        {
            get
            {
                return new[]
                {
                    HookPoint.TurnCounterUpdating, 
                };
            }
        }

        protected override void OnAttachment(ICharacter character)
        {
            var charPAR = character.Stats.GetTotalStat(DebuffResistance.PAR) * 3;
            if (charPAR < 20)
            {
                charPAR = 20;
            }

            var parStrength = character.TimeToAction * (100 / charPAR);
            SetVar(character, "par_Strength", parStrength);
        }

        protected override void OnTurnCounterUpdating(ICharacter character, BoolConsolidator result)
        {
            var timeStep = character.Stats.GetTotalStat(BaseStat.SPD);
            var currentStrength = GetVar(character, "par_Strength");
            var newStrength = currentStrength - timeStep;
            SetVar(character, "par_Strength", newStrength);

            if (newStrength <= 0)
            {
                RemoveFrom(character);
            }
            result.AddResult(false);
        }
    }
}